Refactor Update Last Read

madmax330 opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/Expensify/issues/212871

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console


- In one chrome log into newDot
- In another incognito window log into another account on newDot
- Send messages to the first user
- If this is a new chat make sure when you click on the chat the messages are marked as read 
- If this is an existing chat:
  - If the report is already open, make sure messages are marked as read as they come in
  - If the report is not already open, make sure messages are marked as read when you click on the report


### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [ ] I added steps for local testing in the `Tests` section
    - [ ] I added steps for Staging and/or Production testing in the `QA steps` section
    - [ ] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [x] MacOS / Chrome
    - [ ] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn’t already exist
    - [x] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
</details>

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

See "Tests"

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

This is not a UI change, so I tested on web to show the right network requests being made to the new API commands

#### Web
<!-- Insert screenshots of your changes on the web platform-->

https://user-images.githubusercontent.com/7016692/176709304-f0571205-14e1-4da2-b2fc-08636a3fb8b6.mov

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->

https://user-images.githubusercontent.com/7016692/176921619-1f3152df-b52c-4ec6-9228-4a6fa59712a1.mov



#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->


https://user-images.githubusercontent.com/7016692/176921236-2363f8b1-3410-462d-a9ed-2cdd2fe12081.mov


#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

#### Android
<!-- Insert screenshots of your changes on the Android platform-->

madmax330 committed:
Refactor markasunread

madmax330 committed:
Read newest action

madmax330 committed:
add openreport

madmax330 committed:
Merge branch 'main' into maxence-rrulr
update branch

madmax330 committed:
polish + style

madmax330 committed:
revert package lock

madmax330 commented:
Updated and tested, this is ready for review

marcaaron commented:
Adding @arosiclair as well since he is tagged on the Web-E PR

arosiclair commented:
I'm seeing an issue where occasionally the report is marked as read briefly and then goes back to unread after switching to it. Seems to be caused by `Report_GetHistory` getting chained. 

Will the functionality for Report_GetHistory get refactored into `OpenReport` with another issue or should it be this one?


https://user-images.githubusercontent.com/6833644/176703255-346b2215-9148-48d0-97e1-fd2fbeaa24ac.mov

arosiclair commented:
@madmax330 also you have to flesh out your QA steps (even if its just "same as tests") and screenshots sections

madmax330 commented:
> Will the functionality for Report_GetHistory get refactored into OpenReport with another issue or should it be this one?

Good observation, Report_GetHistory is a separate refactor that will update this behavior. See: https://github.com/Expensify/Expensify/issues/212871#issuecomment-1168063361

> also you have to flesh out your QA steps (even if its just "same as tests")

Good call. 

> and screenshots sections

Not sure what to include for this still since it's not a UI change really, but I'll include a video or the indicator updating

madmax330 commented:
Updated and added a description to the screenshots section

arosiclair commented:
> Not sure what to include for this still since it's not a UI change really, but I'll include a video or the indicator updating and the network tab to show that we're making the requests to the new API

Thanks but you'll have to include recordings for each platform unfortunately since this is UI behavior we're implementing. I'll be testing this on each platform too (minus Android since the build is broken on M1 right now).

arosiclair commented:
> I don't agree with this. These changes don't modify or add any UI components so none of the components will look different than they did before this PR. All it's modifying is the api calls.

I don't like the cross-platform testing requirement for functional changes either, but the guidelines don't really give us an exception. I'll ask for clarification in eng-chat.

madmax330 committed:
Merge branch 'main' into maxence-rrulr
udpate branch

madmax330 commented on src/libs/actions/Report.js:
```diff
@@ -1109,6 +1119,84 @@ function updateLastReadActionID(reportID, sequenceNumber, manuallyMarked = false
+ * @param {Number} reportID
+ */
+function openReport(reportID) {
+    const sequenceNumber = reportMaxSequenceNumbers[reportID];
```
Yeah I thin we should remove this when we implement `OpenReport` b/c we don't need to pass a sequence number to it, it will automatically get it after loading the reportActions and set it to the latest

marcaaron replied:
This is fine for now if we are merging your PR before mine I can make some changes.

madmax330 commented on src/libs/actions/Report.js:
```diff
@@ -1109,6 +1119,84 @@ function updateLastReadActionID(reportID, sequenceNumber, manuallyMarked = false
+                onyxMethod: 'merge',
+                key: `${ONYXKEYS.COLLECTION.REPORT}${reportID}`,
+                value: {
+                    lastVisitedTimestamp: Date.now(),
```
Sure, just, 2 observations, I copied this from `setLocalLastRead` so we've been setting `Date.now()`. 
And this value never reaches the server, but I guess we should make it the same as what the server sends back since the server will be setting this in openReport.

marcaaron replied:
Yep I'm aware that the existing logic is wrong. And still think it's a good opportunity to fix it and not continue to do something that is wrong.

marcaaron replied:
Sorry, I need to mention that I realized I was totally wrong about this. Did some digging and basically...

- report action timestamps are seconds since epoch
- this lastVisitedTimestamp is milliseconds since epoch

madmax330 commented on src/pages/home/report/ContextMenu/ContextMenuActions.js:
```diff
@@ -139,7 +139,7 @@ export default [
         successIcon: Expensicons.Checkmark,
         shouldShow: type => type === CONTEXT_MENU_TYPES.REPORT_ACTION,
         onPress: (closePopover, {reportAction, reportID}) => {
-            Report.updateLastReadActionID(reportID, reportAction.sequenceNumber, true);
```
Good catch I totally missed that, I will add that in

madmax330 replied:
For this particular change, I think we only need to set the `lastReadSequenceNumbers[reportID] = lastReadSequenceNumber; ` because there should always be a sequenceNumber when we're marking as unread. And even if the report is loading we can set it.

Same for the others, I think we only need to set the `lastReadSequenceNumbers` b/c the report shouldn't be loading since we'll be updating with `OpenReport` if the report is loading. And the sequenceNumber should be there for the others.

marcaaron replied:
Ok that sounds right to me as the `unreadActionCount` will be updated (as well as it can be) when we set the `optimisticData`, but we maybe still need to update the `lastReadSequenceNumbers` map with the new value. 

All this code is kind of tricky because the "set local last read" method is more of a "set local last read and reset unread count" (which is totally confusing). I'm drafting something up to get more ideas about how to handle this long term.

marcaaron replied:
Honestly, I kinda want to kill the entire `lastReadSequenceNumbers` entirely if you're down for that 😄 

I started drafting some ideas here that make sense to me, but haven't tested them yet

https://github.com/Expensify/App/pull/9755

madmax330 replied:
Ok so going to set it here, and let you kill it in your PR

madmax330 commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -145,6 +145,7 @@ class ReportActionsView extends React.Component {
             this.updateNewMarkerAndMarkReadOnce();
         }
 
+        Report.openReport(this.props.reportID);
```
I would think that we would remove the call to `updateNewMarkerAndMarkReadOnce()` no? Because `OpenReport` should update the last read since we'll no longer need to wait for the report data to load.

marcaaron replied:
Yeah that sounds correct, but we are still not ready to include this yet because this logic is already handled by the call to `updateNewMarkerAndMarkReadOnce()`

madmax330 commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -353,7 +355,7 @@ class ReportActionsView extends React.Component {
         // Only mark as read if the report is open
         if (!this.props.isDrawerOpen) {
-            Report.updateLastReadActionID(this.props.reportID);
+            Report.readNewestAction(this.props.reportID);
```
As I mentioned in the other comment, I think we can do away with this method once `OpenReport` is implemented

marcaaron replied:
I'm less certain it can be removed as the "new marker" is still getting updated here. But I think the design for that is all wrong and have been looking into how to improve it.

marcaaron replied:
It's a bit more complicated than that because the "marker position" is currently tracked as separate property in Onyx and can also be a different value than the last read.

madmax330 committed:
Address comments

marcaaron commented:
>I'm on latest web and app, I think it's the AddComment PR that broke this (b/c even without my latest changes it's not working). Do I need to be on the latest Auth as well @marcaaron ?

Shouldn't need any Auth changes. Is it broken if you switch to `main`?

marcaaron commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -301,7 +303,7 @@ class ReportActionsView extends React.Component {
     scrollToBottomAndUpdateLastRead() {
         ReportScrollManager.scrollToBottom();
-        Report.updateLastReadActionID(this.props.reportID);
+        Report.readNewestAction(this.props.reportID);
```
yeah, let's 🔪 it and see if anyone misses it?

madmax330 commented:
> Is it broken if you switch to main?

Yeah it's broken on main for me.

madmax330 commented:
> unless you have an idea about what is broken

I don't, but I can look as well 🙃

marcaaron commented:
Ok so doing some debugging and I think it has to do with the timestamp check. For some reason, the timestamp of the actions getting created on dev is off by a solid 30 seconds.

![2022-07-07_08-08-12](https://user-images.githubusercontent.com/32969087/177840591-1aa59daa-4e73-4bd8-a1f9-19c12032b1dc.png)

That means the logic here won't run...

https://github.com/Expensify/App/blob/eb54f232638d6a41db9170de591506d464941850/src/libs/actions/Report.js#L1495-L1498

Seems to work ok on staging though.

marcaaron commented:
Ok, so pretty weird stuff, but I think we are both affected by the VM clock drifting based on this thread -> https://expensify.slack.com/archives/C03TQ48KC/p1657218624480699

After a `vagrant reload` the unreads started to work normally again for me.

madmax330 commented:
Oh wow, you're right, restarting the vm fixed it. I wonder what happened between yesterday and the day before for my clock to skew by 30s 😕

madmax330 commented on src/libs/actions/Report.js:
```diff
@@ -118,6 +118,21 @@ function getUnreadActionCount(report) {
+
+    // Determine the number of unread actions by deducting the last read sequence from the total. If, for some reason,
+    // the last read sequence is higher than the actual last sequence, let's just assume all actions are read
+    return Math.max(reportMaxSequenceNumber - sequenceNumber - ReportActions.getDeletedCommentsCount(reportID, sequenceNumber), 0);
```
Looks like this might be tackled in your other PR to get rid of lastReadSequenceNumber?

marcaaron replied:
Yep yep

madmax330 committed:
Address comments again

madmax330 commented:
Aright, I addressed the comments and re-tested. 
The rest will either be removed in the PR that will implement OpenReport or in your PR to clean up the unread count stuff.
lmk what you think

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -1084,6 +1094,81 @@ function updateLastReadActionID(reportID, sequenceNumber, manuallyMarked = false
+                onyxMethod: 'merge',
+                key: `${ONYXKEYS.COLLECTION.REPORT}${reportID}`,
+                value: {
+                    lastVisitedTimestamp: Math.round(Date.now() / 1000),
```
NAB and sorry, as mentioned in that other comment I was wrong about this and `Date.now()` is totally fine to use as it returns milliseconds and the lastVisitedTimestamp should be in `ms`.

marcaaron reviewed (approved):
Sorry @madmax330 meant to approve this but changes LGTM. Gonna wrap up a few other things in my open PR so you don't get clobbered with conflicts 😅

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

marcaaron commented:
GH actions did pass.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/2694130502) to production by @chiragsalian in version: 1.1.84-13 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
